File: 1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one/1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one.cpp
class Solution {
public:
    int numSteps(string s) {
        int n = s.size(), operations = 0, carry = 0;
        
        for (int i = n - 1; i > 0; --i){
            int digit = (s[i] - 48) + carry; 
            
            if (digit % 2 == 1){
                operations += 2;
                carry = 1;
            }
            else
                ++operations;
        }
        
        return operations + carry;
    }
};

File: 1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one/1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one.py
class Solution:
    def add_one(self, s: List[str]) -> None:
        i = len(s) - 1
        
        # Iterating while the character is 1 and changing to 0
        while i >= 0 and s[i] != '0':
            s[i] = '0'
            i -= 1
        
        if i < 0:
            s.insert(0, "1")
        else:
            s[i] = "1"
    
    def numSteps(self, s: str) -> int:
        s = list(s)
        operations = 0
        
        while len(s) > 1:
            n = len(s)
            
            if s[n - 1] == '0':
                s.pop()
            else:
                self.add_one(s)
        
            operations += 1
        
        return operations